src/sync/sync_user.cpp

```diff
@@ -42,11 +45,21 @@ SyncUser::SyncUser(std::string refresh_token, std::string identity,
         }
     }
     if (token_type == TokenType::Normal) {
-        SyncManager::shared().perform_metadata_update([this, server_url=std::move(server_url)](const auto& manager) {
-            auto metadata = SyncUserMetadata(manager, m_identity);
-            metadata.set_state(server_url, m_refresh_token);
-            m_is_admin = metadata.is_admin();
+        REALM_ASSERT(m_server_url.length() > 0);
+        bool updated = SyncManager::shared().perform_metadata_update([=](const auto& manager) {
+            auto metadata = manager.get_or_make_user_metadata(m_identity, m_server_url);
+            metadata->set_user_token(m_refresh_token);
+            m_is_admin = metadata->is_admin();
+            m_local_identity = metadata->local_uuid();
         });
+        if (!updated)
+            m_local_identity = m_identity;
+    } else {
+        // Admin token users. The local identity serves as the directory path.
+        if (!local_identity)
```

Comment:
Does the absence of a local identity at this point indicate a problem with data that the user has provided, or a failure elsewhere within our code to enforce our preconditions? If the latter, this should be an assertion rather than a exception.